### ISSUE TITLE
Tizen Studio signing broke after Ubuntu 20210831.9 upgrade

### ISSUE THREAD

**Issue by iBotPeaches** [actions/virtual-environments#4108]
### Description

Our builds began failing randomly and while it took a solid amount of time for us to track it down, we identified a regression from the instant the 20210831.9 build went live.

I'm not sure the best method to share a build script that is private, but the tool was Tizen Studio for automating building and signing Tizen applications. 

```
2021-09-08 20:40:38,701 [TRACE] PackageCLI.java(1183) - Execute PackageCLI...
2021-09-08 20:40:38,701 [TRACE] PackageCLI.java(1185) - getRealWorkingPath: /home/runner/work/xxx/xxx/tizen/.buildResult
2021-09-08 20:40:38,891 [ERROR] AbstractCLI.java(93) - java.lang.NullPointerException
java.lang.IllegalStateException: java.lang.NullPointerException
	at org.tizen.common.core.command.policy.UncaughtExceptionHandlingPolicy.uncaughtException(UncaughtExceptionHandlingPolicy.java:84)
	at org.tizen.common.core.command.Executor.execute(Executor.java:179)
	at org.tizen.common.core.command.Executor.execute(Executor.java:150)
	at org.tizen.ncli.subcommands.sign.SignCLICommand.getSigningProfile(SignCLICommand.java:139)
	at org.tizen.ncli.subcommands.sign.SignCLICommand.call(SignCLICommand.java:73)
	at org.tizen.ncli.subcommands.sign.SignCLICommand.call(SignCLICommand.java:58)
	at org.tizen.ncli.subcommands.AbstractSubCommand.runCommand(AbstractSubCommand.java:76)
	at org.tizen.ncli.subcommands.packages.PackageCLICommand.call(PackageCLICommand.java:177)
	at org.tizen.ncli.subcommands.packages.PackageCLICommand.call(PackageCLICommand.java:77)
	at org.tizen.ncli.subcommands.AbstractSubCommand.runCommand(AbstractSubCommand.java:76)
	at org.tizen.ncli.ide.shell.PackageCLI.doPackageForWgt(PackageCLI.java:898)
	at org.tizen.ncli.ide.shell.PackageCLI.execute(PackageCLI.java:1258)
	at org.tizen.ncli.ide.shell.AbstractCLI.execute(AbstractCLI.java:91)
	at org.tizen.ncli.ide.shell.Main.run(Main.java:188)
	at org.tizen.ncli.ide.shell.Main.main(Main.java:121)
Caused by: java.lang.NullPointerException
	at org.tizen.ncli.core.ConsolePrompter.password(ConsolePrompter.java:188)
	at org.tizen.ncli.core.ConsolePrompter.batch(ConsolePrompter.java:262)
	at org.tizen.ncli.core.ConsolePrompter.batch(ConsolePrompter.java:223)
	at org.tizen.ncli.core.ConsolePrompter.batch(ConsolePrompter.java:204)
	at org.tizen.common.sign.command.ReadSigningProfileFileCommand.interactForPassword(ReadSigningProfileFileCommand.java:221)
	at org.tizen.common.sign.command.ReadSigningProfileFileCommand.run(ReadSigningProfileFileCommand.java:100)
	at org.tizen.common.core.command.Executor.execute(Executor.java:164)
	... 13 more
2021-09-08 20:40:38,895 [ERROR] Main.java(196) - java.lang.NullPointerException
```

Since it worked prior on the prior version - 20210816.1, it seems somehow the upgrade broke something that handled password entry and appears to be prompting for an interactive password. I'm just honestly lost on what to do here, our builds are broken. It seems related to an Ubuntu upgrade that I cannot replicate on an update to date Ubuntu machine so then I think Actions is at fault.

### Virtual environments affected

- [ ] Ubuntu 16.04
- [ ] Ubuntu 18.04
- [X] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

```
Virtual Environment
  Environment: ubuntu-20.04
  Version: 20210831.9
```

### Is it regression?

yes - 20210816.1 was last working.

### Expected behavior

The build to be signed and produced.

### Actual behavior

The build fails with an NPE.

### Repro steps

Unfortunately, the repo is private. So I'll do my best to show reproduction steps. Our team is debating internally to rewrite it into a public action so it can be used for easier replication, but for now - this is all I've got.

```
wget http://download.tizen.org/sdk/Installer/tizen-studio_4.1/web-cli_Tizen_Studio_4.1_ubuntu-64.bin
chmod a+x web-cli_Tizen_Studio_4.1_ubuntu-64.bin

TIZEN_STUDIO=$GITHUB_WORKSPACE/tizen-studio
./web-cli_Tizen_Studio_4.1_ubuntu-64.bin --accept-license $TIZEN_STUDIO

tizen build-web -- dist
tizen package -t wgt -s certificiate-name -o tizen -- tizen/.buildResult
```

**Comment by dibir-magomedsaygitov** [comment]
Hello @iBotPeaches. Thank you for your report. We will take a look.

**Comment by dsame** [comment]
@iBotPeaches hi, while we are reproducing the problem with the artificial sample can you please make sure the builds use the same versions of JRE. I noticed the installer brings its own JRE so please double check PATH and JAVA_HOME points to this JRE, not the one pre-installed on actions host 

And would it be possible to get at least source code of one class: org.tizen.ncli.core.ConsolePrompter so we can get some idea about the reason of having NPE

**Comment by iBotPeaches** [comment]
@dsame - Thanks, we will double check the JRE difference as I was not aware the installer provided its own and swap to it in a test. Tizen Studio is open source, but its arguably difficult to follow among the 400 repos.

The IDE repo looks to have the files responsible for that generation if you wish to peek them - https://review.tizen.org/git/?p=sdk/ide/cli.git;a=tree;f=org.tizen.cli/src/org/tizen/cli/exec

In the meantime, we are investigating building a GitHub Action to isolate the behavior and make it easier for replication.

**Comment by dsame** [comment]
@iBotPeaches  

I believe it is just a typo, but while we investigating Tizen, please take a not jdk should be first in the path:

```
PATH="$JAVA_HOME/bin:$PATH"
```

**Comment by iBotPeaches** [comment]
Thanks, good catch @dsame. Unfortunately same issue.

```
Run TIZEN_STUDIO=$GITHUB_WORKSPACE/tizen-studio
  TIZEN_STUDIO=$GITHUB_WORKSPACE/tizen-studio
  PATH="$TIZEN_STUDIO/tools/ide/bin:$PATH"
  JAVA_HOME="$TIZEN_STUDIO/jdk"
  PATH="$JAVA_HOME/bin:$PATH"
  mv dist tizen/
  tizen build-web -- tizen
  tizen package -t wgt -s xxx-o tizen -- tizen/.buildResult
  shell: /usr/bin/bash -e {0}
...
Output path : /home/runner/work/xxx/xxx/tizen/.buildResult
Excluded : [.build/*, .build, .sign/*, .sign, webUnitTest/*, webUnitTest, .externalToolBuilders/*, .externalToolBuilders, .buildResult/*, .buildResult, .settings/*, .settings, .package/*, .package, .tproject, .project, .sdk_delta.info, .rds_delta, *.wgt, .tizen-ui-builder-tool.xml]
Total time: 00:00:00.347
Author password: An error has occurred. See the log file tizen-sdk/tools/ide/cli.log.
Error: Process completed with exit code 1.
```

Compared to the last working build (20ish days ago)
```
Output path : /home/runner/work/xxx/xxx/tizen/.buildResult
Excluded : [.build/*, .build, .sign/*, .sign, webUnitTest/*, webUnitTest, .externalToolBuilders/*, .externalToolBuilders, .buildResult/*, .buildResult, .settings/*, .settings, .package/*, .package, .tproject, .project, .sdk_delta.info, .rds_delta, *.wgt, .tizen-ui-builder-tool.xml]
Total time: 00:00:00.509
Author certficate: /home/runner/work/xxx/xxx/author_cert.p12
Distributor1 certificate : /home/runner/work/xxx/xxx/tizen-studio/tools/certificate-generator/certificates/distributor/sdk-partner/tizen-distributor-signer.p12
Excludes File Pattern: {.manifest.tmp, .delta.lst}
Ignore File: /home/runner/work/xxx/xxx/tizen/.buildResult/.manifest.tmp
Package File Location: /home/runner/work/xxx/xxx/tizen/file.wgt
```

**Comment by dsame** [comment]
@iBotPeaches

I created a project which is close to your repro suggestion and it succeeded, so can you please take a look at my workflow https://github.com/dsame/actions-demo/blob/tizen/.github/workflows/blank.yml and try to modify it in order to make it to fail?

The build https://github.com/dsame/actions-demo/runs/3674532974?check_suite_focus=true

**Comment by iBotPeaches** [comment]
> @iBotPeaches
> 
> I created a project which is close to your repro suggestion and it succeeded, so can you please take a look at my workflow https://github.com/dsame/actions-demo/blob/tizen/.github/workflows/blank.yml and try to modify it in order to make it to fail?
> 
> The build https://github.com/dsame/actions-demo/runs/3674532974?check_suite_focus=true

Hi @dsame , thanks a ton for starting that off. I adapted it until it matched our testing (Just generated a new test certificate that is obviously public and revoked now). However, to my strange-ness it worked.

Our real one continues to fail, so we are entering an internal phase of compare/contract this public one and our own, which is scarily already almost 100% matched, just the difference being GitHub Secrets in place of raw known secrets.

I will follow up, but I opened up a PR for now to match near 100%.

**Comment by dsame** [comment]
@iBotPeaches i merged the PR and will take a look in order to find out the way to make it to fail :)

**Comment by dsame** [comment]
@iBotPeaches also i tryed to look through Tizen source code but i saw only "org.tizen.cli." package while we the problem is "org.tizen.ncli." package. Do you know where are the sources of NCLI command? I believe it will give us an idea about the root problem.

**Comment by dsame** [comment]
@iBotPeaches 
In fact "just the difference being GitHub Secrets in place of raw known secrets" looks suspicious - i guess the pipeline does not get environment variables expected by ncli and this causes the NPE.

Can you please clarify what does it mean "raw known secrets" - environment varibales? Hardcodede ones? File input? Can you make sure "GitHub Secrets" really substitutes (become env.variables, written into files) "raw known secrets"?

**Comment by iBotPeaches** [comment]
> Can you please clarify what does it mean "raw known secrets" - environment varibales? Hardcodede ones? File input? Can you make sure "GitHub Secrets" really substitutes (become env.variables, written into files) "raw known secrets"?

We have two secrets in our real build which is the base64 version of the p12 file and the author password for that certificate. These are loaded into the repository secrets. Like the public version of the build shows now, instead of us building a profile with contents and hard-coded passwords. We just take the secrets from `${{ secrets.foo }}` and add them into the variable expansion to build the profile file with the location to the disassembled b64 certificate as well as the password.

Checks I've done since then
 * The md5 of the reversed b64 p12 from GitHub Secrets matches the md5 from my desktop, so the certificate is getting in the proper place
 * The actual profile file (if I change build script to upload it, since it gets partially redacted) matches what I would expect with the real temporary path to the certificate and the real plaintext password in it.
 * I regenerated another real certificate with a new secure password and still replicates.
 
 I'm now beyond confused. I will dig into the missing `ncli` package you mentioned.
 
 Thanks as always for the help.

**Comment by mspaansen** [comment]
😀

**Comment by iBotPeaches** [comment]
Okay, this is resolved now. I spent probably way too long on this. Long story short - something has the strangest password requirements. It seems that Tizen in comparison with the specific software installed on GHA requires a purely alpha lowercase password (of at least 27 characters).

When we put our real cert/pass into the test action - it failed
When we put the fake cert/pass into the real action - it worked

So this pointed at our certification generation.

So we regenerated the cert again - it failed
So we did it again - it failed
So we did it again w/ the passphrase of the test cert - it worked
So we did it again generating a 27 digital alphanumerical (no special) - it failed
So we did it again generating a 27 alpha (uppercase too) - it failed
So we did it again generating a 27 alpha all lowercase - it worked

How did we get to 27? The example certificate from Tizen uses - `passwordfortestauthorsigner`

At this point exhausting the time/support of too many people. I will just live with this without understanding why and move on.

Thank you for the help @dsame
